luigi/contrib/bigquery.py

```diff
@@ -507,40 +529,155 @@ def run(self):
+                        uris.append(uri)
+                    num_of_uris = len(uris)
+
+                    chunk_intervals, uris_per_chunk = self.calculate_chunk_intervals(gcs_client, uris, output.chunk_size_gb)
+
+                    print "About to schedule " + str(len(chunk_intervals)) + " chunks, each max " \
+                          + str(uris_per_chunk) + " uris"
+
+                    for chunk_num in chunk_intervals:
+                        chunk_uris = uris[chunk_num:min(chunk_num+uris_per_chunk, num_of_uris)]
+                        # min func to handle the very last chunk from the list, usually shorter then "uris_per_chunk"
+                        partial_table_id = "TEMP_" + output.table.table_id + "_" + str(source_counter) + "_" \
+                                           + str(chunk_num) + "_" + str(int(load_start_time))
+
+                        submit_load_job(chunk_uris, partial_table_id)
+                else:
+                    # we don't chunk for non-wildcard source_uri
+                    print "Non-wildcard path for: " + source_uri
+                    table_id = "TEMP_" + output.table.table_id + "_" + str(source_counter) + "_" + str(int(load_start_time))
+                    submit_load_job([source_uri], table_id)
+
+                source_counter += 1
+
+            self.merge_tables(bq_client, output, partial_table_ids, project_id)
+
+            self.remove_tables(bq_client, partial_table_ids)
+
+        load_end_time = time.time()
+        print "UPLOAD DONE, it took: " + str(load_end_time - load_start_time) + " seconds"
+
+    @staticmethod
+    def calculate_avg_blob_size(gcs_client, uris):
+        population_size = min(MAX_POPULATION_SIZE, len(uris))
+        sampled_uris = random.sample(uris, population_size)
+        total_sampled_blobs_bytes = 0
+        for uri in sampled_uris:
+            bucket_name = uri.split("/")[2]
+            blob_name = "/".join(uri.split("/")[3:])
+            blob = gcs_client.get_object(bucket_name, blob_name)
+            if blob is not None:
+                blob_size_bytes = blob['size']
+                total_sampled_blobs_bytes += int(blob_size_bytes)
+            else:
+                raise Exception("BigQueryLoadTask failed: Object + " + uri + " doesn't exist")
+        return total_sampled_blobs_bytes / population_size
+
+    def calculate_chunk_intervals(self, gcs_client, uris, chunk_size_gb):
+        avg_blob_size_bytes = self.calculate_avg_blob_size(gcs_client, uris)
+        #  todo: what if avg_blob_size_bytes * num_of_uris < MAX_UPLOAD_SIZE_BYTES ??
+        #  todo: then we will upload this data in one chunk, but at the end we will still merge this one table
+        #  todo: so this step could be skipped
+        print "Summary avg blob size is: " + str(avg_blob_size_bytes) + " bytes"
+        chunk_size_bytes = chunk_size_gb * GB_TO_BYTES
+        uris_per_chunk = min(MAX_SOURCE_URIS, max(1, chunk_size_bytes / avg_blob_size_bytes))
```

Comment:
:star: chunking on the MAX_SOURCE_URIS limit is value added. Perhaps that's another case in point that we should just only support the "chunked" route - the "non chunked" route misses out on this fix.